streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java

```diff
@@ -154,6 +160,11 @@ private RuntimeException suspendRunningTasks(final Set<TaskId> runningTasksToSus
             } finally {
                 removeTaskFromRunning(task);
                 taskChangelogs.addAll(task.changelogPartitions());
+
+                // Until KAFKA-9177 is fixed, we need to remove from restoredPartitions when a running task is
+                // suspended/revoked instead of when it finishes restoring since the partitions will just be added
+                // back the next time #updateRestored is called
```

Comment:
Cannot follow here -- why `since` ? Would adding partitions not be idempotent?

Reply:
The problem is that `restoredPartitions` was not being cleared/emptied properly. Ideally we would just remove completed partitions from it when they complete in `#updateRestored(restored)`, but the `restored` parameter includes _every_ partition that has finished restoring so far, which we then add back to `restoredPartitions` -- so one `runLoop` later those partitions would be added back, and the only way to remove those partitions again would be to do so when the (now running) task is suspended/closed.
Does that make sense ...?

Reply:
Let me know if there's a better way to phrase the comment as well

Reply:
Maybe:
```
// Until KAFKA-9177 is fixed, we need to remove from restoredPartitions when a running task is
// suspended/revoked instead of when it finishes restoring
// this is save, since the partitions will just be added back the next time #updateRestored is called
```

Reply:
Hm. That seems a bit misleading, since the partitions being added back the next time is actually the _problem_  (can't permanently remove partitions), for which the temporary workaround is to remove the partitions here instead.
It's a mess I know...is this any better
`//Until KAFKA-9177 is fixed, we need to remove from restoredPartitions when a running 
// task is suspended/revoked instead of when it finishes restoring. We cannot permanently 
// remove completed partitions in #updateRestored, since they will be added back to  
// restoredPartitions the next time #updateRestored is called`